azure-iot-device/azure/iot/device/common/async_adapter.py

```diff
@@ -33,20 +36,46 @@ class AwaitableCallback(object):
     """A sync callback whose completion can be waited upon.
     """
 
-    def __init__(self, callback):
-        """Creates an instance of an AwaitableCallback from a callback function.
-
-        :param callback: Callback function to be made awaitable.
+    def __init__(self, return_arg_name=None):
+        """Creates an instance of an AwaitableCallback
         """
+
+        # LBYL because this mistake doesn't cause an exception until the callback
+        # which is much later and very difficult to trace back to here.
+        if return_arg_name and not isinstance(return_arg_name, str):
+            raise TypeError("internal error: return_arg_name must be a string")
+
         loop = asyncio_compat.get_running_loop()
         self.future = asyncio_compat.create_future(loop)
 
         def wrapping_callback(*args, **kwargs):
-            result = callback(*args, **kwargs)
             # Use event loop from outer scope, since the threads it will be used in will not have
-            # an event loop. future.set_result() has to be called in an event loop or it does not work.
-            loop.call_soon_threadsafe(self.future.set_result, result)
-            return result
+            # an event loop. future.set_result() and future.set_exception have to be called in an
+            # event loop or they do not work.
+            if "error" in kwargs and kwargs["error"]:
```

Comment:
Yes, this is a new convention for our project.  Those are good ideas for parameter names but I think this is good for now.  It gives us what we want without extra complexity.

As far as the difference between error and exception, there is something interesting in the language that leads my thinking in a different (actually unknown) direction.  

An exception (lower case e) is a fault that _encompasses_  an Error object (the thing that is raised) along with a traceback.  The traceback is a call stack inside a custom object which cannot be instantiated directly.  The only way to create a traceback object is to raise a BaseException.  The fact that you can't manually create one is strange to me.  You can use `inspect.stack()` to get the current callstack, but that is not a traceback.  I don't know all the history here, but I have read that traceback objects rely heavily on interpreter internals, and I do know that a callstack is "the current stack" but a traceback can be (I assume) a chain of stacks (because of `raise X from Y` functionality).

The reason I chose the word _encompasses_ above instead of _contains_ is because the Error does not contain a pointer to the traceback (like it does in Node).  When an exception is caught, you typically get the traceback from the sys.exc_info() call, which is only documented to work inside an `except:` block.  If you don't retrieve the traceback from inside the `except:` block, it is lost forever.  

All of this together means that (unlike Node), an Error object is not sufficient to contain everything that went wrong.  It contains the "what", but it does not contain the "where".  Being somewhat tainted by my use of Node, I was surprised by this-- the node Error() constructor captures the current call stack and stores it inside an Error object attribute.  My twitch reaction would be to introduce another team convention to add a "callstack" or "traceback" attribute to all of our Error objects, but I am leery of this because a. it's unconventional and b. it introduces garbage collection issues because the traceback contains pointers to stack frames which contain pointers to all stack-local variables.

That being said, in Python 3 (PEP 3109), Exception has an internal __traceback__ attribute which seems like it _could_ be used, but, like all dunders, it's murky.  There's some interesting talk here ([https://stackoverflow.com/a/11415140](https://stackoverflow.com/a/11415140)) that I need to dig into some more.  

U G L Y

---
In reply to: [318838993](https://github.com/Azure/azure-iot-sdk-python-preview/pull/174#discussion_r318838993) [](ancestors = 318838993,318778614)